gencmdref script ending in runtime error

I am trying to move `bin/gencmdref` to the ipfs/website directly, because the Makefile there uses it, it has a hardcoded to path to where it is stored on @jbenet's machine, and there's now been two PRs to the script in this repository to fix an issue in ipfs/website (#1733 and #1723). 

Running `ipfs commands | gencmdref` breaks, however, and I'm not sure why. Here is my exit code: 

```sh
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/ipfs/go-ipfs/commands/cli.parseArgs(0x8210fb4d0, 0x0, 0x3, 0x82111c008, 0xf41a80, 0x2, 0x2, 0x0, 0xf42220, 0x0, ...)
	/go/src/github.com/ipfs/go-ipfs/commands/cli/parse.go:266 +0x10ea
github.com/ipfs/go-ipfs/commands/cli.Parse(0x8210f23d0, 0x3, 0x3, 0x82111c008, 0xf42220, 0x0, 0x0, 0x821249d00, 0x0, 0x0, ...)
	/go/src/github.com/ipfs/go-ipfs/commands/cli/parse.go:45 +0x52b
main.(*cmdInvocation).Parse(0x8210f2840, 0x88212a2110, 0x8210fb440, 0x8210f23d0, 0x3, 0x3, 0x0, 0x0)
	/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:236 +0x89
main.main()
	/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:107 +0x3db

goroutine 18 [chan receive]:
github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics.init.1.func2()
	/go/src/github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics/metrics.go:321 +0x73
created by github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics.init.1
	/go/src/github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics/metrics.go:328 +0x7b

goroutine 19 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37
Traceback (most recent call last):
  File "gencmdref", line 38, in <module>
    main()
  File "gencmdref", line 28, in main
    print run((line + ' --help').split(' ')).strip()
  File "gencmdref", line 11, in run
    return check_output(cmd)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['ipfs', 'dht', 'put', '--help']' returned non-zero exit status 2
```

run it like:
```
ipfs commands | ./gencmdref  # the ./ is important
```

but yeah i see the same error as @RichardLitt  -- after a bunch of good ouput, i get: 

```
## ipfs dht put

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/ipfs/go-ipfs/commands/cli.parseArgs(0x821364d50, 0x0, 0x3, 0x82111a008, 0xf45e40, 0x2, 0x2, 0x0, 0xf465e0, 0x0, ...)
	/go/src/github.com/ipfs/go-ipfs/commands/cli/parse.go:266 +0x10ea
github.com/ipfs/go-ipfs/commands/cli.Parse(0x8210f62d0, 0x3, 0x3, 0x82111a008, 0xf465e0, 0x0, 0x0, 0x821251d00, 0x0, 0x0, ...)
	/go/src/github.com/ipfs/go-ipfs/commands/cli/parse.go:45 +0x52b
main.(*cmdInvocation).Parse(0x821360340, 0x88213220f8, 0x821364cc0, 0x8210f62d0, 0x3, 0x3, 0x0, 0x0)
	/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:236 +0x89
main.main()
	/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:107 +0x3db

goroutine 17 [chan receive]:
github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics.init.1.func2()
	/go/src/github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics/metrics.go:321 +0x73
created by github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics.init.1
	/go/src/github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/codahale/metrics/metrics.go:328 +0x7b

goroutine 18 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37
Traceback (most recent call last):
  File "bin/gencmdref", line 38, in <module>
    main()
  File "bin/gencmdref", line 28, in main
    print run((line + ' --help').split(' ')).strip()
  File "bin/gencmdref", line 11, in run
    return check_output(cmd)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['ipfs', 'dht', 'put', '--help']' returned non-zero exit status 2
```

this is weird because running `ipfs dht put --help` runs fine on its own. even weirder: it breaks on the cli only (i.e. running without a daemon!)

--- a/commands/cli/parse.go
+++ b/commands/cli/parse.go
@@ -297,10 +297,17 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
 
 		var err error
 		if argDef.Type == cmds.ArgString {
-			if stdin == nil || !argDef.SupportsStdin {
+			if stdin == nil {
 				// add string values
 				stringArgs, inputs = appendString(stringArgs, inputs)
+			} else if !argDef.SupportsStdin {
+				if len(inputs) == 0 {
+					// failure case, we have stdin, but our current
+					// argument doesnt want stdin
+					break
+				}
 
+				stringArgs, inputs = appendString(stringArgs, inputs)
 			} else {
 				if len(inputs) > 0 {
 					// don't use stdin if we have inputs

--- a/commands/cli/parse_test.go
+++ b/commands/cli/parse_test.go
@@ -212,7 +212,7 @@ func TestArgumentParsing(t *testing.T) {
 		}
 	}
 
-	testFail := func(cmd words, msg string) {
+	testFail := func(cmd words, fi *os.File, msg string) {
 		_, _, _, err := Parse(cmd, nil, rootCmd)
 		if err == nil {
 			t.Errorf("Should have failed: %v", msg)
@@ -220,18 +220,18 @@ func TestArgumentParsing(t *testing.T) {
 	}
 
 	test([]string{"noarg"}, nil, []string{})
-	testFail([]string{"noarg", "value!"}, "provided an arg, but command didn't define any")
+	testFail([]string{"noarg", "value!"}, nil, "provided an arg, but command didn't define any")
 
 	test([]string{"onearg", "value!"}, nil, []string{"value!"})
-	testFail([]string{"onearg"}, "didn't provide any args, arg is required")
+	testFail([]string{"onearg"}, nil, "didn't provide any args, arg is required")
 
 	test([]string{"twoargs", "value1", "value2"}, nil, []string{"value1", "value2"})
-	testFail([]string{"twoargs", "value!"}, "only provided 1 arg, needs 2")
-	testFail([]string{"twoargs"}, "didn't provide any args, 2 required")
+	testFail([]string{"twoargs", "value!"}, nil, "only provided 1 arg, needs 2")
+	testFail([]string{"twoargs"}, nil, "didn't provide any args, 2 required")
 
 	test([]string{"variadic", "value!"}, nil, []string{"value!"})
 	test([]string{"variadic", "value1", "value2", "value3"}, nil, []string{"value1", "value2", "value3"})
-	testFail([]string{"variadic"}, "didn't provide any args, 1 required")
+	testFail([]string{"variadic"}, nil, "didn't provide any args, 1 required")
 
 	test([]string{"optional", "value!"}, nil, []string{"value!"})
 	test([]string{"optional"}, nil, []string{})
@@ -239,14 +239,14 @@ func TestArgumentParsing(t *testing.T) {
 
 	test([]string{"optionalsecond", "value!"}, nil, []string{"value!"})
 	test([]string{"optionalsecond", "value1", "value2"}, nil, []string{"value1", "value2"})
-	testFail([]string{"optionalsecond"}, "didn't provide any args, 1 required")
-	testFail([]string{"optionalsecond", "value1", "value2", "value3"}, "provided too many args, takes 2 maximum")
+	testFail([]string{"optionalsecond"}, nil, "didn't provide any args, 1 required")
+	testFail([]string{"optionalsecond", "value1", "value2", "value3"}, nil, "provided too many args, takes 2 maximum")
 
 	test([]string{"reversedoptional", "value1", "value2"}, nil, []string{"value1", "value2"})
 	test([]string{"reversedoptional", "value!"}, nil, []string{"value!"})
 
-	testFail([]string{"reversedoptional"}, "didn't provide any args, 1 required")
-	testFail([]string{"reversedoptional", "value1", "value2", "value3"}, "provided too many args, only takes 1")
+	testFail([]string{"reversedoptional"}, nil, "didn't provide any args, 1 required")
+	testFail([]string{"reversedoptional", "value1", "value2", "value3"}, nil, "provided too many args, only takes 1")
 
 	// Use a temp file to simulate stdin
 	fileToSimulateStdin := func(t *testing.T, content string) *os.File {
@@ -296,6 +296,7 @@ func TestArgumentParsing(t *testing.T) {
 	fstdin = fileToSimulateStdin(t, "stdin1")
 	test([]string{"stdinenablednotvariadic2args", "value1"}, fstdin, []string{"value1", "stdin1"})
 	test([]string{"stdinenablednotvariadic2args", "value1", "value2"}, fstdin, []string{"value1", "value2"})
+	testFail([]string{"stdinenablednotvariadic2args"}, fstdin, "cant use stdin for non stdin arg")
 
 	fstdin = fileToSimulateStdin(t, "stdin1")
 	test([]string{"noarg"}, fstdin, []string{})

--- a/core/commands/dht.go
+++ b/core/commands/dht.go
@@ -12,8 +12,8 @@ import (
 	notif "github.com/ipfs/go-ipfs/notifications"
 	path "github.com/ipfs/go-ipfs/path"
 	ipdht "github.com/ipfs/go-ipfs/routing/dht"
-	u "gx/ipfs/QmZNVWh8LLjAavuQ2JXuFmuYH3C11xo988vSgp7UQrTRj1/go-ipfs-util"
 	peer "gx/ipfs/QmUBogf4nUefBjmYjn6jfsfPJRkmDGSeMhNj4usRKq69f4/go-libp2p/p2p/peer"
+	u "gx/ipfs/QmZNVWh8LLjAavuQ2JXuFmuYH3C11xo988vSgp7UQrTRj1/go-ipfs-util"
 )
 
 var ErrNotDHT = errors.New("routing service is not a DHT")

